api/src/pcapi/alembic/versions/20230504T124540_1a27ac9f25a2_create_trusted_device.py

```diff
@@ -0,0 +1,34 @@
+"""create trusted device table
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# pre/post deployment: pre
+# revision identifiers, used by Alembic.
+revision = "1a27ac9f25a2"
+down_revision = "b1ee65265887"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    op.create_table(
+        "trusted_device",
+        sa.Column("id", sa.BigInteger(), autoincrement=True, nullable=False),
+        sa.Column("userId", sa.BigInteger(), nullable=False),
+        sa.Column("deviceId", sa.String(), nullable=False),
+        sa.Column("source", sa.String(), nullable=True),
+        sa.Column("os", sa.String(), nullable=True),
+        sa.Column("dateCreated", sa.DateTime(), nullable=False),
+        sa.ForeignKeyConstraint(["userId"], ["user.id"], ondelete="CASCADE"),
```

Comment:
**(problème / downtime)** J'ai un peu peur de cette clé étrangère. La création de la table devra prendre un lock sur la table `user`. Même si la création sera instantanée, le lock lui-même peut être problématique. Si on a une requête `UPDATE` longue sur la table `user`, le ShareRowExclusiveLock demandé par `CREATE TABLE` va "s'empiler" à la suite, et toutes les requêtes `UPDATE` sur la table `user` s'empileront derrière l'`ALTER TABLE`, bloquant toute modification sur la table (mais pas les lectures). On n'a normalement pas de requête `UPDATE` longue sur la table `user`, donc le risque est sûrement relativement faible... mais l'impact serait très important.

Mais je ne vois pas de contournement. On pourrait créer la table sans la clé étrangère et l'ajouter ensuite avec `NOT VALID` : 

```sql
alter table "trusted_devices"
    add constraint "xxx_fkey" foreign key ("userId") references "user" (id) not valid;
```

... prend également un ShareRowExclusiveLock sur la table "user", donc ça ne résout pas notre problème. :) Je crois qu'on n'a pas vraiment de solution magique, à part (1) espérer ne pas avoir de requête `UPDATE` longue sur "user" ; et (2) avoir un "lock timeout" court pour ne pas attendre trop longtemps et bloquer trop longtemps les autres requêtes `UPDATE` s'empilant derrière le `CREATE TABLE`.

**Il me semble que le "lock timeout" est actuellement à 0 (zéro), donc on attend indéfiniment, ce qui est problématique pour ce genre de migration.** Je suggère donc si possible de ne **pas merger** cette pull request avant d'avoir correctement configuré le "lock timeout". Il me semble que le sujet a été discuté très récemment par @francois-seguin et @jroussel-pass .

Reply:
Alembic [utilise](https://github.com/pass-culture/pass-culture-main/blob/2e1563750786bf97713c1779b8c232c05390c69e/api/src/pcapi/alembic/run_migrations.py#L34-L35C1) la valeur de `DB_MIGRATION_STATEMENT_TIMEOUT`, qui est actuellement sa valeur par défaut, `5000` ms.

Reply:
Ah ! Merci de cette correction, François. Puisque la valeur est relativement courte, je clos cette conversation en "résolue".